subsys/bootloader/bl_storage/bl_storage_nvmc.c

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright (c) 2024 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <bl_storage.h>
+#include <bl_storage_internal.h>
+#include <zephyr/kernel.h>
+
+/** Function for getting the current value and the first free slot.
+ *
+ * @param[in]   counter_desc  Counter description
+ * @param[out]  counter_value The current value of the requested counter.
+ * @param[out]  free_slot     Pointer to the first free slot. Can be NULL.
+ *
+ * @retval 0        Success
+ * @retval -EINVAL  Cannot find counters with description @p counter_desc or the pointer to
+ *                  @p counter_value is NULL.
+ */
+int get_counter(uint16_t counter_desc, uint16_t *counter_value, const uint16_t **free_slot)
+{
+	const uint16_t *slots;
+	uint16_t highest_counter = 0;
+	const uint16_t *addr = NULL;
+	const struct monotonic_counter *counter_obj = get_counter_struct(counter_desc);
+	uint16_t num_counter_slots;
+
+	if (counter_obj == NULL || counter_value == NULL) {
+		return -EINVAL;
+	}
+
+	slots = counter_obj->counter_slots;
+	num_counter_slots = bl_storage_otp_halfword_read((uint32_t)&counter_obj->num_counter_slots);
+
+	for (uint32_t i = 0; i < num_counter_slots; i++) {
+		uint16_t counter = ~nrfx_nvmc_otp_halfword_read((uint32_t)&slots[i]);
+
+		if (counter == 0) {
+			addr = &slots[i];
+			break;
+		}
+		if (highest_counter < counter) {
+			highest_counter = counter;
+		}
+	}
+
+	if (free_slot != NULL) {
+		*free_slot = addr;
+	}
+
+	*counter_value = highest_counter;
+	return 0;
+}
+
+int get_monotonic_counter(uint16_t counter_desc, uint16_t *counter_value)
```

Comment:
we could probably save on duplicate code by defining counter type and reusing the `get/set` functions with proper counter type definitions.